src/app/gateways/views/GatewayPoliciesView.vue

```diff
@@ -19,33 +19,76 @@
-              :src="`/meshes/${route.params.mesh}/gateways/${route.params.dataPlane}/policies`"
+              v-slot="{ data: policyTypesData, error: policyTypesError }: PolicyTypeCollectionSource"
+              :src="`/*/policy-types`"
             >
-              <ErrorBlock
-                v-if="policyTypesError"
-                :error="policyTypesError"
-              />
-
-              <ErrorBlock
-                v-else-if="error"
-                :error="error"
-              />
-
-              <LoadingBlock v-else-if="data === undefined || policyTypesData === undefined" />
-
-              <GatewayDataplanePolicyList
-                v-else
-                :policy-types-by-name="policyTypesData.policies.reduce((obj, policyType) => Object.assign(obj, { [policyType.name]: policyType }), {})"
-                :gateway-dataplane="data"
-              />
+              <DataSource
+                v-slot="{ data: gatewayDataplane, error }: MeshGatewayDataplaneSource"
+                :src="`/meshes/${route.params.mesh}/gateways/${route.params.dataPlane}/policies`"
+              >
+                <ErrorBlock
+                  v-if="policyTypesError"
+                  :error="policyTypesError"
+                />
+
+                <ErrorBlock
+                  v-else-if="error"
+                  :error="error"
+                />
+
+                <LoadingBlock v-else-if="gatewayDataplane === undefined || policyTypesData === undefined" />
+
+                <GatewayDataplanePolicyList
+                  v-else
+                  :policy-types-by-name="policyTypesData.policies.reduce((obj, policyType) => Object.assign(obj, { [policyType.name]: policyType }), {})"
+                  :gateway-dataplane="gatewayDataplane"
+                />
+              </DataSource>
+            </DataSource>
+          </template>
+
+          <template v-else>
+            <DataSource
+              v-slot="{ data: policyTypesData, error: policyTypesError }: PolicyTypeCollectionSource"
+              :src="`/*/policy-types`"
+            >
+              <DataSource
+                v-slot="{ data: sidecarDataplane, error }: SidecarDataplaneCollectionSource"
+                :src="`/meshes/${route.params.mesh}/dataplanes/${route.params.dataPlane}/sidecar-dataplanes-policies`"
+              >
+                <DataSource
+                  v-slot="{ data: rulesData, error: rulesError }: DataplaneRulesCollectionSource"
+                  :src="`/meshes/${route.params.mesh}/dataplanes/${route.params.dataPlane}/rules`"
+                >
+                  <ErrorBlock
+                    v-if="policyTypesError"
+                    :error="policyTypesError"
+                  />
+
+                  <ErrorBlock
+                    v-else-if="error"
+                    :error="error"
+                  />
+
+                  <ErrorBlock
+                    v-else-if="rulesError"
+                    :error="rulesError"
+                  />
```

Comment:
Little nit, I 'think' or-ing would work here also 

```html
<ErrorBlock
  v-if="policyTypesError || error || rulesError"
  :error="policyTypesError || error || rulesError"
/>
```

As long as the `v-if/:error` are in the same order the correct error would be passed to the `ErrorBlock` I think.

FYI: I have a first draft `DataLoader` sitting here almost ready to PR, which hopefully will help with all this type of stuff everywhere in the application.

---

I just saw theres another one further up where we use 2 ErrorBlocks also.

These are a non-blocker btw, up to you if you prefer it (and I'm correct that it would be fine)

Reply:
This is actually necessary because those two variants are not equivalent. The type system doesn’t know that after a check `(T1 | undefined) || (T2 | undefined)` it’s safe to assign to a value expecting `T`.